backend/antibiootit/src/main/java/fi/tuni/koodimankelit/antibiootit/builder/DiagnoseResponseBuilder.java

```diff
@@ -0,0 +1,161 @@
+
+        }
+    };
+
+
+
+    /**
+     * Default constructor
+     * @param diagnose Database entity instance
+     * @param weight weight in kilograms
+     * @param usePenicillinAllergic True, if penicillin allergic option should be used
+     */
+    public DiagnoseResponseBuilder(Diagnose diagnose, double weight, boolean usePenicillinAllergic) {
+        this.diagnose = diagnose;
+        this.weight = weight;
+        this.usePenicillinAllergic = usePenicillinAllergic;
+    }
+
+    
+    /** 
+     * Build diagnose response object
+     * @return DiagnoseResponse generated instance
+     */
+    public DiagnoseResponse build() {
+
+        DiagnoseResponse diagnoseResponse = new DiagnoseResponse(diagnose.getId(), diagnose.getEtiology());
+        List<Treatment> treatments = getTreatments();
+
+        for(Treatment treatment : treatments) {
+            Antibiotic antibiotic = getSuitableAntibiotic(treatment);
+
+            AntibioticTreatmentBuilder builder = new AntibioticTreatmentBuilder(antibiotic, this.weight);
+            AntibioticTreatment antibioticTreatment = builder.build();
+
+            diagnoseResponse.addTreatment(antibioticTreatment);
+        }
+
+        return diagnoseResponse;
+        
+    }
+
+    
+    /** 
+     * Returns all suitable treatments based on penicillin allergy
+     * @return List<Treatment> suitable treatments
+     */
+    private List<Treatment> getTreatments() {
+
+        List<Treatment> treatments = new ArrayList<>();
+        for(Treatment treatment : this.diagnose.getTreatments()) {
+            if(isSuitableTreatment(treatment)) {
+                treatments.add(treatment);
+            }
+        }
+
+        // Sort by treatment choice: primary, secondary, penicillinAllergic
+        treatments.sort((a, b) -> Integer.valueOf(a.getChoice()).compareTo(Integer.valueOf(b.getChoice())));
+        return treatments;
+        
+    }
+
+    
+    /** 
+     * Return True if treatment is suitable based on penicillin allergy
+     * @param treatment specific treatment
+     * @return boolean True, if treatment is suitable
+     * @throws NoAntibioticTreatmentException if diagnose has no antibiotic treatment
+     */
+    private boolean isSuitableTreatment(Treatment treatment) {
+        if(this.usePenicillinAllergic) {
+            return PENICILLIN_ALLERGIC_CHOICE == treatment.getChoice();
+        }
+        else if (NO_ANTIBIOTIC_CHOICE == treatment.getChoice()) {
+            throw new NoAntibioticTreatmentException(this.diagnose);
+        }
+        else {
+            return PRIMARY_CHOICE == treatment.getChoice() || SECONDARY_CHOICE == treatment.getChoice();
+        }
+    }
+
+    
+    /** 
+     * Returns suitable antibiotic choise from given treatment based on weight
+     * @param treatment specific treatment
+     * @return Antibiotic preferred antibiotic
+     */
+    private Antibiotic getSuitableAntibiotic(Treatment treatment) {
```

Comment:
Here's the assumption I mentioned. A treatment can have multiple different antibiotics, and those can have multiple different strengths. These strengths have their minimum weights. For example these could be:

Antibiotic A
- minWeight 0

Antibiotic B
- minWeight 40
- minWeight 60

This function sorts minWeight to descending order within an antibiotic. Then it compares the antibiotics, which one has the higher minWeight is first. Then the function iterates the antibiotics and their strengths and chooses the strength with highest applicable minWeight.

This assumes, that different antibiotics don't have overlapping minWeight strengths. For example:

Antibiotic C
- minWeight 0
- minWeight 50

Antibiotic D
- minWeight 40
- MinWeight 60

They would be ordered:
60 - D
40 - D
50 - C
0   - C

If our patient is 55kg, the 40 - D is chosen, even though 50 - C is the most suitable one. I think the antibiotics' strength's minimum weights don't overlap like this?